### ISSUE TITLE
Issue with not rolled back transactions

### ISSUE THREAD

**Issue by Melzar** [odavid/typeorm-transactional-cls-hooked#43]
Hello,

First of all thanks for your work.

I have an issue with not rolled back transactions in my repository, even I can clearly see that in DB logs transaction is starting and rolling back. This issue occurs in my mocha tests ( data are not cleared between tests ). 

I've following configuration steps but the issue still occurs. Any hints regarding what I'm doing wrong?

To reproduce just update line `115` in `Equipment.spec.ts` file to `test equipment`

Link to repository

https://github.com/Melzar/onion-architecture-boilerplate

Thanks in advance for any help and your time.

**Comment by odavid** [comment]
Thank you @Melzar!
Looking at your test code, I do not see you **did not** [initialize](https://github.com/odavid/typeorm-transactional-cls-hooked#initialization) the cls namespace.

You should do it before these test run:
```typescript
import { initializeTransactionalContext } from 'typeorm-transactional-cls-hooked';

initializeTransactionalContext() // Initialize cls-hooked
``` 

Hope it helps...

**Comment by Melzar** [comment]
@odavid Thank you for your answer. 

I know what you mean but:

1. I'm doing it in `prepareTestDB` function ( at the top of the file there is a helper )
2. Even when I move `initializeTransactionalContext` at the top of the test file ( or before hook  - ) then I have the following error 

` Error: No CLS namespace defined in your app ... please call initializeTransactionalContext() before application start.
`
I remember I've managed to work around this issue somehow but still transaction rollback wasn't applied ( but I could see rollbacks and so on in DB logs )

I would be grateful for any hints.

**Comment by odavid** [comment]
Hi @Melzar 
Took your code in order to investigate. 
* First comment: ormconfig.sample.js should be changed - paths are `.../db/...` instead of `.../database/...` - typeorm could not load entities.
* initializeTransactionalContext should be the first thing in the file, since it initializes the cls-hooked context.

I am sorry, but I could not find why (spent 2 hours...)... 
Have you tried to talk with @Kerren-Entrostat about that? Maybe he can help here...
Please let me know if you managed to solve this...

**Comment by odavid** [comment]
Hi @Melzar, 
I think I have a clue...
It seems that you change the connection name during the tests. so it starts with onion_test and then becomes /v1/equipment and then again onion_test. I believe this is the main issue, since the context name is built on the connection name.

Hope it helps...

**Comment by Melzar** [comment]
@odavid Regarding paths - I've refactored code structure so that was the reason - it requires just pull from repository as I've updated dependencies there.

Connection name is the same during the tests but I'm creating new database for every test suite, so tests can run in parallel. Did you manage to debug this issue with changed connection name during tests? I haven't ask @Kerren-Entrostat about that, I'm not sure but I've got a feeling that this issue can be caused by AbstractRepository in TypeOrm which I'm extending. I will try to debug source code of your lib and see what is the issue. Maybe somewhere along the code lines rollback are applied not to the DB with provided name but to default database.


I'll let you know about my findings.

**Comment by Melzar** [comment]
@odavid I think I know what is going on, I'll try to explain.

The issue probably is related to @Transactional decorator, as it takes as parameter connectionName which for which we have then EntityManager. `InTransaction` hook in my repo, is evaluated and set when code is compiled and then `onion_test` database options are picked from ormconfig.js file. It looks like Transaction is applied to `onion_test` connection and entityManager not `/v1/equipment` or any other connections and databases. That's why probably data from databases are not cleared - transaction is not applied to correct database. To solve it, I would need a way to pass custom connection data / connection / entityManager to transaction. Do you have any idea how can I do it? If extension is needed to your lib I'll prepare something, but probably I will have to prepare additional wrapper function with optional parameters as it won't work probably with decorator - what do you think?

**Comment by odavid** [comment]
@Melzar - maybe connectionName can be either a string or a function that provides string...
I am +- AFK and won't be able to test that. Can you check that and if yes prepare a PR?

**Comment by Melzar** [comment]
@odavid I've checked again - Test passed when I've provided connectionName '/v1/equipment' and commented out rest of the tests. I'll see if there is workaround for that and provide you some feedback

**Comment by odavid** [comment]
@Melzar - Can u please try the PR I've created here. and pass connectionName: ()=> process.env.ORM_CONNECTION in your code?

**Comment by Melzar** [comment]
@odavid I'm on it, let me check

**Comment by odavid** [comment]
Please take it again... had a small bug...

**Comment by Melzar** [comment]
@odavid Checking... I have some issues with resolving modules but it's probably on my side

**Comment by Melzar** [comment]
I can't test it because sources are in TS not js and I have following issue

`export enum IsolationLevel {
^^^^^^

SyntaxError: Unexpected token export
`
Any idea to workaround it?

**Comment by odavid** [comment]
please run `yarn build` and then use the dist folder...

**Comment by Melzar** [comment]
```
Unhandled asynchronous exception: TypeError: Object.defineProperty called on non-object
    at Function.defineProperty (<anonymous>)
    at onion-architecture-boilerplate/dist/node_modules/typeorm-transactional-cls-hooked/src/Transactional.js:25:16
```
Hmm not sure if it's on my side or bug

**Comment by Melzar** [comment]
Ok I've double checked your solution - it won't work because it's a decorator which is evaluated before everything else :( so before I'm able to mock/stub function ( I have to do it in every spec in before call ) we are still returning 'onion_test' value. I think that without transaction wrapper function it won't be possible to achieve dynamic values.

**Comment by odavid** [comment]
@Melzar - I made it work just now.
Please pull and try again...

* Commented out `// await clearTestDB(testName);` in `Equiment.spec.js` so db won't deleted
* Ran psql
```
postgres=# \c /v1/equipment
psql (12.1, server 11.6 (Debian 11.6-1.pgdg90+1))
You are now connected to database "/v1/equipment" as user "postgres".
/v1/equipment=# select * from equipment;
 id | name | userId 
----+------+--------
(0 rows)

/v1/equipment=# ```

Before it had a value all the time.

**Comment by Melzar** [comment]
@odavid Ok checking...

**Comment by odavid** [comment]
@Melzar - looking forward for results 😄

**Comment by Melzar** [comment]
@odavid You are the man! It works - but, there is one eventual issue, you can see it also in my repo and is related to User.spec.ts. We are running tests in transactions and if in repository we try to remove records in transaction ( or apply any operation in transaction, like adding / removing / updating related records ) we will face following error `Error: Cannot start transaction because its already started` - do you have any idea to workaround this issue ?

**Comment by odavid** [comment]
Well, I would remove the from the repository and move it to the service level method by wrapping it with @Transactional 

It is too late for me... will merge it soon and create a new release...

**Comment by Melzar** [comment]
Got it, I'll play around and see what I can do, thanks man for all the support!

**Comment by odavid** [comment]
@Melzar - released 0.1.12
